mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaDAO.java

```diff
@@ -0,0 +1,119 @@
+        entityManager = entityManagerFactory.createEntityManager();
+    }
+
+    public void setMaxStorage(QuotaRoot quotaRoot, Optional<QuotaSize> maxStorageQuota) {
+        entityManager.getTransaction().begin();
+        entityManager.merge(new MaxUserStorage(quotaRoot.getValue(), quotaValueToLong(maxStorageQuota)));
+        entityManager.getTransaction().commit();
+    }
+
+    public void setMaxMessage(QuotaRoot quotaRoot, Optional<QuotaCount> maxMessageCount) {
+        entityManager.getTransaction().begin();
+        entityManager.merge(new MaxUserMessageCount(quotaRoot.getValue(), quotaValueToLong(maxMessageCount)));
+        entityManager.getTransaction().commit();
+    }
+
+    public void setDefaultMaxStorage(Optional<QuotaSize> defaultMaxStorage) {
+        entityManager.getTransaction().begin();
+        entityManager.merge(new MaxDefaultStorage(quotaValueToLong(defaultMaxStorage)));
+        entityManager.getTransaction().commit();
+    }
+
+    public void setDefaultMaxMessage(Optional<QuotaCount> defaultMaxMessageCount) {
+        entityManager.getTransaction().begin();
+        entityManager.merge(new MaxDefaultMessageCount(quotaValueToLong(defaultMaxMessageCount)));
+        entityManager.getTransaction().commit();
+    }
+
+    public Optional<QuotaSize> getDefaultMaxStorage() {
+        MaxDefaultStorage storedValue = entityManager.find(MaxDefaultStorage.class, MaxDefaultStorage.DEFAULT_KEY);
+        return longToQuotaSize(storedValue.getValue());
+    }
+
+    public Optional<QuotaCount> getDefaultMaxMessage() {
+        MaxDefaultMessageCount storedValue = entityManager.find(MaxDefaultMessageCount.class, MaxDefaultMessageCount.DEFAULT_KEY);
+        return longToQuotaCount(storedValue.getValue());
+    }
+
+    public Optional<QuotaSize> getMaxStorage(QuotaRoot quotaRoot) {
+        MaxUserStorage storedValue = entityManager.find(MaxUserStorage.class, quotaRoot.getValue());
+        return longToQuotaSize(storedValue.getValue());
+    }
+
+    public Optional<QuotaCount> getMaxMessage(QuotaRoot quotaRoot) {
+        MaxUserMessageCount storedValue = entityManager.find(MaxUserMessageCount.class, quotaRoot.getValue());
+        return longToQuotaCount(storedValue.getValue());
+    }
+
+
+    private Long quotaValueToLong(Optional<? extends QuotaValue<?>> maxStorageQuota) {
+        return maxStorageQuota.map(value -> {
+            if (value.isUnlimited()) {
+                return INFINITE;
+            }
+            return value.asLong();
+        }).orElse(null);
+    }
+
+    private Optional<QuotaSize> longToQuotaSize(Long value) {
+        return longToQuotaValue(value, QuotaSize.unlimited(), QuotaSize::size);
+    }
+
+    private Optional<QuotaCount> longToQuotaCount(Long value) {
+        return longToQuotaValue(value, QuotaCount.unlimited(), QuotaCount::count);
+    }
+
+    private <T extends QuotaValue<T>> Optional<T> longToQuotaValue(Long value, T infiniteValue, Function<Long, T> quotaFactory) {
```

Comment:
Reminds me something.... I read this in the Cassandra implementation. Maybe it more belongs to QuotaValue?

Reply:
It does not. Encoding values is strictly a backend specific responsibility. Making this code common between implementation is not a very good idea because it would bound two backends more than needed.